src/main/resources/google_checks.xml

```diff
@@ -126,7 +126,8 @@
                     LITERAL_FOR, LITERAL_IF, LITERAL_RETURN, LITERAL_SWITCH, LITERAL_SYNCHRONIZED,
                     LITERAL_TRY, LITERAL_WHILE, LOR, LT, MINUS, MINUS_ASSIGN, MOD, MOD_ASSIGN,
                     NOT_EQUAL, PLUS, PLUS_ASSIGN, QUESTION, RCURLY, SL, SLIST, SL_ASSIGN, SR,
-                    SR_ASSIGN, STAR, STAR_ASSIGN, LITERAL_ASSERT, TYPE_EXTENSION_AND"/>
+                    SR_ASSIGN, STAR, STAR_ASSIGN, LITERAL_ASSERT,
+                    TYPE_EXTENSION_AND, LITERAL_WHEN"/>
```

Comment:
@nrmancuso @romani ping.

@mahfouz72 If we are changing google checks, we need documentation that the google style guide supports this change.
Basically what specific rule supports this.
https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html

For our config, we like to use as much as possible, but only if we really agree with the rule being applied to our repo. If we don't agree with it, we usually still add it but set it's violations to be ignored so we are still atleast no exception testing it.

Reply:
@Zopsss ping as well

Reply:
[4.6.2 Horizontal whitespace](https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s4.6.2-horizontal-whitespace) section defines rules for Whitespace around code. The documentation does not specifically mentions anything about the `when` keyword but the google-java-format adds whitespace around `when`. Here's the formatted code:

```java
record ColoredPoint(int p, int x, String c) {}

record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) {}

public class Test {

  void test(Object obj) {
    int y = 0;
    if (y >= 2) {} // violation
    switch (obj) {
      case ColoredPoint(int x, _, _) when (x >= 2) -> {} // expected violation
      default -> {}
    }
  }

  void test2(Object obj) {
    int y = 0; // 2 violations, no space before and after '='
    switch (obj) {
      case ColoredPoint(int x, _, _) when (x
          >= 2) -> {} // (expected) 2 violations, no space before and after 'when'
      case ColoredPoint(int x, _, _) when (x
          >= 0) -> {} // (expected) 1 violation, no space after 'when'
      default -> {}
    }
  }
}
```

`when` was introduced in Java 21, the last major update for style guide was before Java 21 released so it does not contains rule for newly introduced keywords. 

But I think we should add support for this keyword, it is ensures basic spacing between `when` and other code and it is overall improving the coding style

Reply:
If we agree with this keyword then I think it's we should test cases for it in [InputWhitespaceAroundBasic.java](https://github.com/checkstyle/checkstyle/blob/master/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputWhitespaceAroundBasic.java) file.

Reply:
@Zopsss Even though it is not specifically mentioned, are there any parts of the style which indicate this spacing follows the heart of the guide? Unfortunately we can't go by what "looks good" since that is subjective.

This is the closest I am seeing
> Separating any reserved word

`when` is a keyword like the others
https://docs.oracle.com/javase/specs/jls/se22/html/jls-3.html#jls-3.9
But it isn't require to have a `(` after it, but the google documentation also clearly states the spaces should only appear in very specific places.

@romani @nrmancuso What do you think?

Reply:
> But it isn't require to have a `(` after it

I see. If that's the case then based on style guide, we don't need any kind rule for `when` as it does not mentions anything explicitly about it. 

Though the formatter does adds spacing around it and I also think we should add support for this token to maintain consistent spacing. I'll be waiting for other maintainers opinion on this.

Reply:
> Separating any reserved word, such as if, for or catch, from an open parenthesis (() that follows it on that line

`when` is a reserved word and it may have an open parenthesis so I think google style should support this token

> But it isn't require to have a ( after it

why does this information affect our decision? Yes it is not required but It still may have an open parenthesis

Reply:
@mahfouz72 Please read https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s4.6.2-horizontal-whitespace .
Google config follows style guide and style guide seems to be very strict in this case on what is supposed to happen with spacing.

Reply:
@mahfouz72 , please use formatter https://github.com/google/google-java-format on input to confirm that `when` in spaced on both sides.

Reply:
@Zopsss used the formatter and confirmed that it adds spaces around `when` https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1686824725

Reply:
> @mahfouz72 , please use formatter https://github.com/google/google-java-format on input to confirm that `when` in spaced on both sides.

Formatter adds spaces on both sides of `when` but there's no spacing rule defined for Java 21 keywords in the style guide because style guide is more older than the Java 21 release.

Reply:
@mahfouz72 , please add example of "when" to new Input at folder
https://github.com/checkstyle/checkstyle/tree/master/src/it/resources-noncompilable/com/google/checkstyle/test/chapter4formatting
and create test method at https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/HorizontalWhitespaceTest.java

Reply:
https://github.com/checkstyle/checkstyle/pull/15359

Reply:
Same state as other PR. @nrmancuso @romani

Reply:
Yes, this PR is above my expectations.
Thanks a lot.